src/v/security/krb5.h

```diff
@@ -10,82 +10,100 @@
 
 #pragma once
 
+#include "outcome.h"
+#include "seastarx.h"
+
+#include <seastar/core/sstring.hh>
+
 #include <krb5/krb5.h>
 
+#include <optional>
 #include <string_view>
 
 namespace security::krb5 {
 
-class krb5_context_view {
+class context_view {
 public:
-    explicit krb5_context_view(const ::krb5_context* ctx)
+    explicit context_view(::krb5_context ctx)
       : _ctx(ctx) {}
 
-    ::krb5_context operator*() const { return *_ctx; }
-    explicit operator const ::krb5_context*() { return _ctx; }
+    ::krb5_context operator*() const { return _ctx; }
 
 private:
-    const ::krb5_context* _ctx;
+    ::krb5_context _ctx;
 };
 
-class krb5_context {
-public:
-    krb5_context() = default;
-    krb5_context(krb5_context&&) = delete;
-    krb5_context(const krb5_context&) = delete;
-    krb5_context operator=(krb5_context&&) = delete;
-    krb5_context operator=(const krb5_context&) = delete;
-
-    ~krb5_context() noexcept { reset(); }
-
-    void reset() noexcept {
-        if (_ctx != nullptr) {
-            ::krb5_free_context(_ctx);
-            _ctx = nullptr;
+namespace impl {
+class error_message {
+    struct deleter {
+        std::optional<context_view> ctx;
+        void operator()(const char* realm) const {
+            ::krb5_free_error_message(ctx ? **ctx : nullptr, realm);
         }
-    }
+    };
 
-    explicit operator krb5_context_view() const {
-        return krb5_context_view{&_ctx};
-    }
-    explicit operator ::krb5_context() const { return _ctx; }
-    ::krb5_context* operator&() { return &_ctx; }
+public:
+    explicit error_message(const char* msg)
+      : _msg(msg, deleter{}) {}
+
+    explicit error_message(const char* msg, context_view ctx)
+      : _msg(msg, deleter{ctx}) {}
+
+    explicit operator std::string_view() { return _msg.get(); }
 
 private:
-    ::krb5_context _ctx{nullptr};
+    std::unique_ptr<const char, deleter> _msg;
+};
+
+struct error_impl {
+    ::krb5_error_code ec{0};
+    ss::sstring msg;
+
+    friend std::ostream&
+    operator<<(std::ostream& os, const struct error_impl& impl) {
+        os << impl.msg << "(ec: " << impl.ec << ")";
+        return os;
+    }
 };
 
-class krb5_default_realm {
+} // namespace impl
+
+template<typename R>
+using result = result<R, impl::error_impl>;
+
+class context {
 public:
-    explicit krb5_default_realm(krb5_context_view ctx)
-      : _ctx(std::move(ctx)) {}
+    static result<context> initialize_krb5_context() noexcept;
 
-    krb5_default_realm(krb5_default_realm&&) = delete;
-    krb5_default_realm(const krb5_default_realm&) = delete;
-    krb5_default_realm& operator=(krb5_default_realm&&) = delete;
-    krb5_default_realm& operator=(const krb5_default_realm&) = delete;
+    explicit operator context_view() { return context_view{_ctx.get()}; }
+    explicit operator ::krb5_context() const { return _ctx.get(); }
```

Comment:
There's a similar problem here; context owns the `::_krb5_context` but because we're dishing out a `::krb5_context`, we're not const correct. The way I see it, the const should propagate through the pointer. 

For interaction with the C api, it might just be clearer to hand out non-const underlying via a member that has a more explicit name.